packages/utilities/src/classNamesFunction.ts

```diff
@@ -1,32 +1,73 @@
-import { mergeStyleSets, IStyle, IStyleSet, IProcessedStyleSet } from '@uifabric/merge-styles';
+import { mergeStyleSets, IStyleSet, IProcessedStyleSet } from '@uifabric/merge-styles';
 import { IStyleFunctionOrObject } from '@uifabric/merge-styles';
 
+const MAX_CACHE_COUNT = 30;
+
+// Note that because of the caching nature within the classNames memoization,
+// I've disabled this rule to simply be able to work with any types.
+// tslint:disable:no-any
+
+const RetVal = '__retval__';
+
 /**
  * Creates a getClassNames function which calls getStyles given the props, and injects them
  * into mergeStyleSets.
+ *
+ * Note that the props you pass in on every render should be in the same order and
+ * immutable (numbers, strings, and booleans). This will allow the results to be memoized. Violating
+ * these will cause extra recalcs to occur.
  */
 export function classNamesFunction<TStyleProps extends {}, TStyleSet extends IStyleSet<TStyleSet>>(): (
   getStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
   styleProps?: TStyleProps
 ) => IProcessedStyleSet<TStyleSet> {
-  // TODO: memoize.
+  let map: Map<any, any> = new Map<any, any>();
 
   const getClassNames = (
     styleFunctionOrObject: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
     styleProps: TStyleProps = {} as TStyleProps
   ): IProcessedStyleSet<TStyleSet> => {
-    // styleSet might be undefined if styleFunctionOrObject is undefined, but getStyles should never
-    // ordinarily be undefined (it would hardly make any sense).
-    // However, because we usually use `props.styles` as the argument to an invocation of this method, and
-    // `props.styles` itself is defined as optional, this avoids the need to use `!` at all invocation points.
-    if (styleFunctionOrObject === undefined) {
-      return {} as IProcessedStyleSet<TStyleSet>;
+    let current: Map<any, any> = map;
+    let resultCount = 0;
+
+    for (const propName in styleProps) {
+      if (styleProps.hasOwnProperty(propName)) {
+        let propValue: any = styleProps[propName];
+
+        if (propValue === undefined) {
+          propValue = '__undefined__';
+        }
+        if (propValue === null) {
+          propValue = '__null__';
+        }
+        if (!current.has(propValue)) {
+          current = current.set(propValue, new Map<any, any>());
+        }
+        current = current.get(propValue);
+      }
     }
 
-    const styleSet =
-      styleFunctionOrObject && (typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps!) : styleFunctionOrObject);
+    if (!(current as any)[RetVal]) {
```

Comment:
I don't understand why do current[RetVal].. why not do `!current.has(RetVal)`? Am I missing something here

Reply:
Not sure if it is the same in TS/JS, but isn't it normally recommended to simply get and check `undefined | null` vs. `has` re: performance? Thought that was the case in Java / C#... would have to measure.

Reply:
I added comments, but the `[RetVal]` is an attached property, not a key in the Set. It represents the return value for the branch in the graph.